### PR TITLE
 vdhcoapp: init at 2.0.19 

### LINKED ISSUE #112046: RFP: Video DownloadHelper Companion App (vdhcoapp)
**Project description**

Video DownloadHelper is a "Browser extension to download videos from the Web"

> Some operations required by Video DownloadHelper cannot be performed from within the browser. In order to be able to still do the job, the add-on relies on an external application that is called transparently. We call it the Companion Application.

**Metadata**

* homepage URL: https://www.downloadhelper.net/install-coapp
* source URL: https://github.com/mi-g/vdhcoapp
* license: gpl2
* platforms: linux, darwin, (windows)

**other build scripts**

* [arch linux](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=vdhcoapp)

### PR DESCRIPTION
**Pull request by HannesGitH**
**fixes #112046** 

adds [vdhcoapp](https://github.com/aclap-dev/vdhcoapp) as a source-built package to pkgs

---

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md). (or at least i hope so)

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/vd/vdhcoapp/ffmpeg-filepicker.patch
Comment (lolbinarycat) [inline pkgs/by-name/vd/vdhcoapp/ffmpeg-filepicker.patch:7]: what's the motivation for making open a noop?

File: pkgs/by-name/vd/vdhcoapp/filepicker.nix
Comment (lolbinarycat) [inline pkgs/by-name/vd/vdhcoapp/filepicker.nix:19]:
```suggestion

  src = fetchFromGitHub {
    owner = "paulrouget";
    repo = "static-filepicker";
    rev = "v${version}";
    hash = "sha256-7sRzf3SA9RSBf4O36olXgka8c6Bufdb0qsuTofVe55s=";
```

needless indirection.

File: pkgs/by-name/vd/vdhcoapp/filepicker.nix
Comment (lolbinarycat) [inline pkgs/by-name/vd/vdhcoapp/filepicker.nix:39]:
```suggestion
    maintainers = with maintainers; [ hannesgith ];
```

comment is unnecessary, the author and maintainer of a package are usually different people.
the author handles the upstream code, and the maintainer handles writing a nix derivation that compiles that code.

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (lolbinarycat) [inline pkgs/by-name/vd/vdhcoapp/package.nix:15]:
```suggestion
```

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (lolbinarycat) [inline pkgs/by-name/vd/vdhcoapp/package.nix:62]:
```suggestion
    runHook preInstall
    mkdir -p $out/opt/vdhcoapp

    cp -r . "$out/opt/vdhcoapp"

    makeWrapper ${nodejs}/bin/node $out/bin/vdhcoapp \
      --add-flags $out/opt/vdhcoapp/src/main.js
    runHook postInstall
```

run hooks manually since we are overriding the installPhase.

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [inline pkgs/by-name/vd/vdhcoapp/package.nix:62]: How to we install the native hosts files, so that we can use it in combination with the browser? Afaik you will have to install them manually atm, which does not feel very user friendly for nixos users

**Comment by Shawn8901** [comment]
Can you please also fix the history and the commit message so that it fits the contribution guidelines?

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [inline pkgs/by-name/vd/vdhcoapp/package.nix:68]: If i understood @wolfangaukang correctly he is not willing to maintain that package in nixpkgs, so i assume that listing here as maintainer is a false claim, isnt it?

Comment (HannesGitH) [reply pkgs/by-name/vd/vdhcoapp/package.nix]: you are right, my thought was that this code was mainly his contribution and he added himself as maintainer in the code he provided i did not deem it fair to remove him, but as he explicity wished to not maintain this package you are totally right in removing him from the maintainers list

File: pkgs/by-name/vd/vdhcoapp/ffmpeg-filepicker.patch
Comment (HannesGitH) [reply pkgs/by-name/vd/vdhcoapp/ffmpeg-filepicker.patch]: as the import statement could not be used this was simply a quick fix, forgotten to be removed, it is now replaced with `const open = import('open')`

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (HannesGitH) [reply pkgs/by-name/vd/vdhcoapp/package.nix:62]: true, this is now done in the postInstall hook. fixed.

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [reply pkgs/by-name/vd/vdhcoapp/package.nix]: Yeah fair claim, maybe he responds and gives context (if he wants to be added).
As the source is his personal repo, I assume that he adds himself as maintainer for own packages (which make sense when using some evals that depend on meta.maintainers, or just wanting to fill it completely). I am doing similar on my personal flake.

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [inline pkgs/by-name/vd/vdhcoapp/package.nix:8]: Why is that callPackage needed?

Comment (HannesGitH) [reply pkgs/by-name/vd/vdhcoapp/package.nix:8]: because filepicker is not in nixpkgs and i chose to not add it as its only used by vdhcoapp

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [reply pkgs/by-name/vd/vdhcoapp/package.nix:62]: Did you try to verify the result?
I am quite sure that running the vdhco binaries installer does not do what we expect from it in the nixpkgs context.

I see that the following is happening

```
vdhcoapp> Writing /build/.mozilla/native-messaging-hosts/net.downloadhelper.coapp.json
vdhcoapp> Writing /build/.var/app/org.mozilla.firefox/.mozilla/native-messaging-hosts/net.downloadhelper.coapp.json
```

So it seems that its installing to `$HOME`, which is not valid in our context here as we are  running from a building user that has its `$HOME` set to `/build`.
We could install them system wide, but they are sadly also not valid in out context, so we might want to patch the file to write to `$out`. 
https://github.com/aclap-dev/vdhcoapp/blob/0b40d3e101c429fdae4e02ba1f593c2416b51eff/config.toml#L187 

In addition it seems to have some installation detection (so for which browser it should install the manifests), and i am not sure if that logic is valid in our context here. I would expect the derivation to have all *common* manifests avaiable (so chrome, chromium, firefox, ...), it we need to provide some way that its detecting in which browser context it is installed to generate the right manifest files.

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [reply pkgs/by-name/vd/vdhcoapp/package.nix:8]: i am not sure if "only used by vdhcoapp" should be the requirement to not expose that as general library, but i definetly see that it might not be very useful atm for others, so i guess that could be still okay. Resolved from my side

**Comment by HannesGitH** [comment]
sorry for asking but as this is my first PR to nixpkgs, how long is ofborg supposed to take for this one missing pipeline? 
(it is now active for nearly 5h)

**Comment by Shawn8901** [comment]
Not sure if you had noticed, but I don't think that my manifest topic has been addressed. 
https://github.com/NixOS/nixpkgs/pull/300051#discussion_r1545344144
If we need to fix that for chrome or so idk, but at least Firefox should be worth as there is an option for that on nixos

**Comment by HannesGitH** [comment]
> Not sure if you had noticed, but I don't think that my manifest topic has been addressed. [#300051 (comment)](https://github.com/NixOS/nixpkgs/pull/300051#discussion_r1545344144) If we need to fix that for chrome or so idk, but at least Firefox should be worth as there is an option for that on nixos.

oh, no i have not (not been notified), i probably thought the script worked because i had run the installer manually beforehand so the files remained in the correct location, which had nothing to do with the installationPhase, my Bad, thanks for noticing!  i will have a look at how to fix that later

> Also the newly added commits are not kind of worth keeping. 

wdym, i should squash the commits into a single one, i.e. not worth keeping in git history, or i should revert/reset them, i.e. not worth keeping the contents of 'em?

> Instead of merging you should also consider rebasing.

aight, i just used the webbased resolver here in github. will also fix that later

**Comment by Shawn8901** [comment]
> wdym, i should squash the commits into a single one, i.e. not worth keeping in git history, or i should revert/reset them, i.e. not worth keeping the contents of 'em?

I would squash the commits together. So keeping a commit that you fixed manifest install is possibly nice for in-between longer reviews, but it's nice to clean up before merging

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (HannesGitH) [reply pkgs/by-name/vd/vdhcoapp/package.nix:62]: > Instead of patching we could also let it write to `$HOME` and move the manifest afterwards to some place, where it would be picked up by the `nativeMessagingHosts` option(s) (hint: idk how that is done for chrome/chromium, but there is an [option](https://search.nixos.org/options?channel=unstable&show=programs.firefox.nativeMessagingHosts.packages&from=0&size=50&sort=relevance&type=packages&query=programs.firefox.nativeMessagingHosts.packages) for at least firefox in nixpkgs and there is also [one in home-manager](https://home-manager-options.extranix.com/?query=programs.firefox.nativeMessagingHosts&release=master), and it would be cool to support those for a good user experience) .

might be kind of a newbee question, but is there a way i could write to those options via my derivation or is the user supposed to do something in the lines of 
```nix
programs.firefox.nativeMassagingHosts = [ "${pkgs.vdhcoapp}/host"]
```
(provided i/this derivation moved the manifest there)

Comment (HannesGitH) [reply pkgs/by-name/vd/vdhcoapp/package.nix:62]: > so we might want to patch the file to write to `$out`

this would still not move them to the correct $HOME, right as $out is /nix/store/bla ?

is there a nice way to access the correct user home (e.g. given as input / mkOption), that would be my ansatz

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (WolfangAukang) [inline pkgs/by-name/vd/vdhcoapp/package.nix:13]:
```suggestion
```
Unnecessary line

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [reply pkgs/by-name/vd/vdhcoapp/package.nix:62]: you can not write to $HOME of *your* user when building a package derivation, which comes from a build server
I mean we are adding it to nixpkgs, so that it gets shipped by nixos binary cache etc arent we? :slightly_smiling_face: 
But luckly there is a nixos option, which will pick up system wide installed manifests.

`programs.firefox.nativeMessagingHosts.packages`  will look for messaging hosts put at `$out/lib/mozilla/native-messaging-hosts/` of the given packages, via the firefox [wrapper](https://github.com/NixOS/nixpkgs/blob/4c147d4246aa82bd105bd805a6209e4a5071dd32/pkgs/applications/networking/browsers/firefox/wrapper.nix#L38).

**Comment by WolfangAukang** [comment]
Small observation: You will need to patch part of VDHCoApp code if it is built for darwin, this line specifically:

https://github.com/aclap-dev/vdhcoapp/blob/0b40d3e101c429fdae4e02ba1f593c2416b51eff/app/src/native-autoinstall.js#L34

Now, I only see `obb` as a package that contains it, but it is marked as broken. You could limit the platforms to only Linux for the moment.

**Comment by HannesGitH** [comment]
> Small observation: You might need to patch part of VDHCoApp code if it is built for darwin, this line specifically:
> 
> https://github.com/aclap-dev/vdhcoapp/blob/0b40d3e101c429fdae4e02ba1f593c2416b51eff/app/src/native-autoinstall.js#L34
> 
> Now, I only see `obb` as a package that contains it, but it is marked as broken. You could limit the platforms to only Linux for the moment.

unfortunately i dont have a darwin system, so its hard to test for me, if you are confident that the darwin variant would be broken, i'd disable it

**Comment by ToasterUwU** [comment]
If there is doubt and nobody will test it soon, I think it's better to just take it out for now and maybe make this it's own seperate issue if anything.
Delaying a package multiple people are waiting for, for an edge case being potentially broken doesn't seem like the right move.

**Comment by Shawn8901** [comment]
not sure if installing the native messaging host is an edge case, as its literally a requirement to use the app in combination with the browsers. but feel free to merge it if you feel that its still a edge condition.

**Comment by ToasterUwU** [comment]
> not sure if installing the native messaging host is an edge case, as its literally a requirement to use the app in combination with the browsers. but feel free to merge it if you feel that its still a edge condition. 
> 
> And for darwin it was suggested how to handle that.

I wasn't talking about the messaging thing.
I just meant the Darwin thing. Nobody seems to be around to test it, so ignore for it now.

**Comment by Shawn8901** [comment]
That's the reason why it was suggested to mark it as not supported / limit supported platform to linux.

**Comment by ToasterUwU** [comment]
> That's the reason why it was suggested to mark it as not supported / limit supported platform to linux for now, unless someone later on is willing to test/fix that.

Yup, exactly what I meant as well.

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [reply pkgs/by-name/vd/vdhcoapp/package.nix:61]: I have a version that builds the needed manifest files adhoc by itself from the config.toml (or the converted json) adhoc on bulid here https://github.com/Shawn8901/nix-configuration/blob/main/packages/vdhcoapp/default.nix

The problem with your approach is that this results in a strange behavior for hydra builds as the auto-install.js code does look up what browsers are actually installed on the system, thus hydra will likely never place manifests in the store.

I tested the firefox variant with that.

No saying that the code is nice, but possibly it could be a starting point for you

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [inline pkgs/by-name/vd/vdhcoapp/package.nix:73]:
```suggestion
    license = licenses.gpl2;
    platforms = lib.platforms.linux;
```
as suggested in https://github.com/NixOS/nixpkgs/pull/300051#issuecomment-2030078101

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (HannesGitH) [reply pkgs/by-name/vd/vdhcoapp/package.nix:61]: Thats a very nice starting point, thanks!

there is no 'nix-way' to add manifest to user installed browsers tho, right?

i think skipping the auto-install and going with your version will be the best way then.

Comment (Shawn8901) [reply pkgs/by-name/vd/vdhcoapp/package.nix:61]: I already referenced nix ways of adding using the manifest when provided by the package in my prior comment https://github.com/NixOS/nixpkgs/pull/300051#discussion_r1545344144 

> Instead of patching we could also let it write to $HOME and move the manifest afterwards to some place, where it would be picked up by the nativeMessagingHosts option(s) (hint: idk how that is done for chrome/chromium, but there is an [option](https://search.nixos.org/options?channel=unstable&show=programs.firefox.nativeMessagingHosts.packages&from=0&size=50&sort=relevance&type=packages&query=programs.firefox.nativeMessagingHosts.packages) for at least firefox in nixpkgs and there is also [one in home-manager](https://home-manager-options.extranix.com/?query=programs.firefox.nativeMessagingHosts&release=master), and it would be cool to support those for a good user experience) .

which contains links to the firefox versions.
For chrome/ium i am not aware of builtins for generically adding packages, tho there is a starting point in adding them hardcoded https://github.com/NixOS/nixpkgs/blob/efb96bc312fdd1fc4fdca67b4901f1a3409b7bda/nixos/modules/programs/chromium.nix#L127

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [inline pkgs/by-name/vd/vdhcoapp/package.nix:67]: this will be likely run on hydra for caching on the binary cache, not sure if it helps the user to dig thru the hydra logs to find that.

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [inline pkgs/by-name/vd/vdhcoapp/package.nix:66]: like elaborated in my comment here https://github.com/NixOS/nixpkgs/pull/300051#discussion_r1564635579 we can not rely on running `vdhcoapp install` as it will break reproducibility and hydra will likely not have a firefox running, thus not create a manifest for firefox.

Comment (HannesGitH) [reply pkgs/by-name/vd/vdhcoapp/package.nix]: yep i understood :) thanks for the guidance

**Comment by Shawn8901** [comment]
i guess we could now squash the commits together (so that one is adding you to the maintainers and then the other for init vdhcoapp), then i will cherry pick that PR to my nixpkgs fork and report pack if it works on my setup (if it works ill approve from my end  :slightly_smiling_face: )

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [inline pkgs/by-name/vd/vdhcoapp/package.nix:29]:
```suggestion
  nativeBuildInputs = [
    jq
```

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [inline pkgs/by-name/vd/vdhcoapp/package.nix:4]:
```suggestion
, jq
, toml2json
```

**Review by Shawn8901 — APPROVED** [review]
works fine for me

If others test that, you should be possible be aware that firefox (or the extension) seems to resolve the wrapper path and show the node path in UI as installed binary, instead the created vdhcoapp wrapper, tho it should work fine in finding the app.

**Comment by ToasterUwU** [comment]
@HannesGitH Im pretty new, so maybe i missed something, but when i test it with nixpkgs-review, the companion app is not detected by default. I have to run `vdhcoapp install` by hand to register it. 
Maybe that is happening automatically if its installed in a nixos config, but if not thats something that should be changed before merging this. Users shouldnt have to run a command by hand to have the software they installed be setup properly.

**Comment by HannesGitH** [comment]
> @HannesGitH Im pretty new, so maybe i missed something, but when i test it with nixpkgs-review, the companion app is not detected by default. I have to run `vdhcoapp install` by hand to register it. Maybe that is happening automatically if its installed in a nixos config, but if not thats something that should be changed before merging this. Users shouldnt have to run a command by hand to have the software they installed be setup properly.

Unfortunately running vdhcoapp install from within the derivation is not really possible as Shawn explained [here](https://github.com/NixOS/nixpkgs/pull/300051#discussion_r1564635579) but if it's installed via the config, and your browser (either Firefox chrome(ium) or edge) is as well, it should work out of the box

For non standard installs I think there is no easy way. The manifest will be placed at `${pkgs.vdhcoapp}/lib/mozilla/native-massaging-hosts` which could be added to the native massaging hosts in your non standard Firefox fork for example (in its config). (That would be the nix/declarative way, otherwise on non nixos with otherwise managed standard browsers, running vdhcoapp install seems like the easiest way)

**Review by ToasterUwU — APPROVED** [review]

**Comment by HannesGitH** [comment]
is there anything else I need to do, or is this just waiting for a maintainer with write access to pick it up?

**Comment by Shawn8901** [comment]
> is there anything else I need to do, or is this just waiting for a maintainer with write access to pick it up?

Generally people with commit access to pick up prs that have approval labels. If it's not happening for some days it's fair to post that on eG. Discourse to get some attention. That what I did now in hope that we get it merged. It would be cool to make it to 24.05 (still some days until that 🙂 )

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (SuperSandro2000) [reply pkgs/by-name/vd/vdhcoapp/package.nix:7]: This will fall apart as soon as another program is added to nixpkgs named filepicker which I am not sure if that will be caught in review.

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/vd/vdhcoapp/package.nix:23]:
```suggestion

  src = fetchFromGitHub {
    owner = "aclap-dev";
    repo = "vdhcoapp";
    rev = "v${version}";
    hash = "sha256-8xeZvqpRq71aShVogiwlVD3gQoPGseNOmz5E3KbsZxU=";
```
if you need to mark that special for an updater script or so, please use an comment instead

File: pkgs/by-name/vd/vdhcoapp/package.nix
Comment (Shawn8901) [reply pkgs/by-name/vd/vdhcoapp/package.nix:7]: yeah moving it to a let statement or directly adding it to nixpkgs might be a better idea.

**Pull request merged**
